### DISCUSSION TITLE
Migrating packages to *Async and away from Fibers

### DISCUSSION THREAD

**Discussion by StorytellerCZ** [Meteor-Community-Packages/organization#68] [General]
With Meteor 2.8 released the time has come to start migrating MCP packages away from the Fibers land as well. As such I would like to discuss which packages to do first and also document here what the progress should be. Since good patterns will emerge over time I would like to start with the smaller packages first.

## Status board
https://github.com/orgs/Meteor-Community-Packages/projects/8/views/1

## Versioning pattern
If only internal APIs change there should be a feature version bump with `api.versionsFrom(['2.8'])` be set. If publicly consumable APIs change then I think we should just do a major version bump.

The README should then include a description which versions are compatible with what versions of Meteor and CHANGELOG should also be very explicit about those things.

Some packages will depend on Meteor API changes that will come in later version ie. Accounts in Meteor 2.9. When that time comes the relevant code should be migrated.

## *Async
Until Meteor 3.0 there should be option to use both synchronous and *Async versions of calls. This in order to give a chance for people to migrate. Where possible *Async function calls should be provided alongside the existing API. Where possible internally switch should be made to use the new *Async API.

### Collection2 - 2.8
I think we can start working on this with release 2.8

### linked-accounts - 2.9
Accounts are scheduled for 2.9 so we have to wait will 2.9 release before releasing this, that said hopefully this shouldn't be too much work.

** TODO **

**Comment by guncebektas** [comment]
I think I will be able to help for some of the packages. We can starts with collection2

**Reply by StorytellerCZ** [reply]
Thank you! Ping me when you have a PR or want to discuss on how to approach this.

**Comment by harryadel** [comment]
Please do so, we need every with MCP. It's too much for @storyteller to
handle right now.

**Comment by perbergland** [comment]
Here are some packages I use that need rewriting and where it’s unclear if there are any active maintainers.

I have created issues for migration to see if there is any response:

littledata:synced-cron - https://github.com/percolatestudio/meteor-synced-cron/issues/147

xolvio:cleaner - https://github.com/xolvio/cleaner/issues/26

percolate:migrations - https://github.com/percolatestudio/meteor-migrations/issues/84

If there is no response I can help moving them to meteor-community and try to fix them

**Reply by StorytellerCZ** [reply]
I think I should be able to publish `percolate:migrations`. For the rest I'm not sure.

**Reply by dr-dimitru** [reply]
@perbergland we at @veliovgroup maintain a good replacement to `synced-cron` — the [`josk` npm package](https://www.npmjs.com/package/josk), we recommend to use NPM version of the package within Meteor setup. 

For cases when NPM package causes an error or conflict we suggest to use [`packosphere` version of the package](https://github.com/veliovgroup/josk/blob/master/docs/meteor.md#josk-usage-within-meteorjs)

Hope that helps

**Reply by StorytellerCZ** [reply]
OK, I can only manage `percolate:migrations` on GitHub, but can't publish. I will negotiate with Meteor Software to see if we can take over this one.

**Reply by Grubba27** [reply]
We were discussing this internally, and maybe if the community sees fit, we can merge some of these packages to the core to boost their migration and give them more attention.
Those that we were more interested in are:
- migrations
- cron
- hooks(hooks for MongoDB, we are still deciding which one)

**Reply by StorytellerCZ** [reply]
To be more specific:
percolate:migrations
percolate:synced-cron
method and collection hooks?

**Reply by fredmaiaarantes** [reply]
The original package candidates are:

```
littledata:synced-cron
percolate:migrations
matb33:collection-hooks
```

As the community is already migrating some of those packages to Meteor 3.0, when we get there, we will look again at the community packages to confirm if those are the most stable or well-crafted for us to use as a base package.

Of course, if the community has other suggestions. We are open to them.

**Reply by StorytellerCZ** [reply]
I think if collection hooks are added, so should be method hooks, or potentially the idea here is to add hooks to existing Meteor methods.

**Comment by minhna** [comment]
alanning:roles

**Reply by ricaragao** [reply]
@minhna , are you getting the meteor:roles to update? If you don't, I will.

**Reply by minhna** [reply]
Yes, please do it. Thank you.
